tests/benchmarks/bench_assembly.cpp

```diff
@@ -0,0 +1,389 @@
+      if (assembly == AssemblyLevel::ELEMENT)
+      {
+         mem += num_elems * pow(p+1, 2*dim) * 8;
+      }
+      if (assembly == AssemblyLevel::FULL)
+      {
+         mem += 3 * num_elems * pow(p+1, 2*dim) * 8;
+      }
+      // std::cout << "mem = " << mem << " , max_mem = " << max_mem << std::endl;
+      return mem < max_mem;
+   }
+
+   virtual void setup() = 0;
+
+   virtual void benchmark() = 0;
+
+   double SumMdofs() const { return mdofs; }
+
+   double MDofs() const { return 1e-6 * dofs; }
+};
+
+/// Bake-off Problems (BPs)
+template<typename BFI, int VDIM = 1, bool GLL = false>
+struct Problem: public BakeOff
+{
+   const double rtol = 1e-12;
+   const int max_it = 32;
+   const int print_lvl = -1;
+
+   Array<int> ess_tdof_list;
+   Array<int> ess_bdr;
+   LinearForm b;
+   OperatorPtr A;
+   Vector B, X;
+   CGSolver cg;
+
+   Problem(AssemblyLevel assembly, int order, int N):
+      BakeOff(assembly,order,N,VDIM,GLL),
+      ess_bdr(mesh.bdr_attributes.Max()),
+      b(&fes)
+   {
+      if (is_runnable())
+      {
+         ess_bdr = 1;
+         fes.GetEssentialTrueDofs(ess_bdr,ess_tdof_list);
+         b.AddDomainIntegrator(new DomainLFIntegrator(one));
+         b.Assemble();
+
+         a.SetAssemblyLevel(assembly);
+         a.AddDomainIntegrator(new BFI(one, GLL?irGLL:ir));
+         a.Assemble();
+         a.Mult(x, y);
+
+         a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);
+
+         cg.SetRelTol(rtol);
+         cg.SetOperator(*A);
+         cg.SetMaxIter(max_it);
+         cg.SetPrintLevel(print_lvl);
+         cg.iterative_mode = false;
+         MFEM_DEVICE_SYNC;
+      }
+   }
+
+   void setup() override
+   {
+      a.Assemble();
+      MFEM_DEVICE_SYNC;
+      mdofs += MDofs();
+   }
+
+   void benchmark() override
+   {
+      cg.Mult(B,X);
+      MFEM_DEVICE_SYNC;
+      mdofs += MDofs() * cg.GetNumIterations();
+   }
+};
+
+/// Bake-off Problems (BPs)
+#define BakeOff_Problem(assembly,i,Kernel,VDIM,p_eq_q)\
+static void SetupBP##i##assembly(bm::State &state){\
+   const int dim = 3;\
+   const int p = state.range(1);\
+   const int target_dofs = state.range(0);\
+   const int elem_dofs = pow(p+1, dim);\
+   const int N = pow(target_dofs / elem_dofs, 1.0/dim) + 1;\
+   Problem<Kernel##Integrator,VDIM,p_eq_q> ker(AssemblyLevel::assembly, p, N);\
+   if ( !ker.is_runnable() ) { state.SkipWithError("MAX_MEM"); }\
+   while (state.KeepRunning()) { ker.setup(); }\
+   state.counters["MDof/s"] = bm::Counter(ker.SumMdofs(), bm::Counter::kIsRate);\
+   state.counters["Dofs"] = bm::Counter(ker.dofs, bm::Counter::kDefaults);\
+   state.counters["Order"] = bm::Counter(ker.p);\
+   state.counters["Assembly"] = (int)AssemblyLevel::assembly;}\
+BENCHMARK(SetupBP##i##assembly)->ArgsProduct({\
+      benchmark::CreateRange(1024, max_dofs, /*step=*/2),\
+      benchmark::CreateDenseRange(1, max_order, /*step=*/1)\
+    })->Unit(bm::kMillisecond);\
+static void BP##i##assembly(bm::State &state){\
```

Comment:
This benchmark seems redundant with the `bench_ceed` already present in the directory.

Reply:
I don't think so, `bench_ceed` is a much smaller subset of these benchmarks. This benchmark is ment to output benchmark information that is almost directly usable in a publication for performance analysis.